# 项目：评估和清理英国电商公司销售数据

## 一、分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 二、简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 三、读取数据

In [8]:
import pandas as pd
import numpy as np
original_data=pd.read_csv("e_commerce.csv")

original_data.rename(columns={"InvoiceNo":"发票号码","StockCode":"产品代码","Description":"产品名称","Quantity":"交易数量","InvoiceDate":"发票日期和时间",
               "UnitPrice":"单价(英镑（£）)","CustomerID":"客户编号","Country":"国家名称"},inplace=True)
original_data.sample(20)

发票号码    产品代码                                产品名称  交易数量  \
304958  563614   22736         RIBBON REEL MAKING SNOWMEN    120   
473138  576780   22784               LANTERN CREAM GAZEBO      9   
163296  550542   21454                PAINTED PINK RABBIT      4   
426805  573364   21591        COSY HOUR CIGAR BOX MATCHES      1   
124629  546984   20754     RETROSPOT RED WASHING UP GLOVES     1   
255958  559418   22976    CIRCUS PARADE CHILDRENS EGG CUP     12   
299525  563104   23321         SMALL WHITE HEART OF WICKER    12   
74754   542526   85062  PEARL CRYSTAL PUMPKIN T-LIGHT HLDR     6   
479354  577156   23287             RED VINTAGE SPOT BEAKER     8   
316247  564752   23312         VINTAGE CHRISTMAS GIFT SACK     4   
84886   543450   20725             LUNCH BAG RED RETROSPOT     1   
32991   539214   22904           CALENDAR PAPER CUT DESIGN     1   
195741  553739   47566                       PARTY BUNTING     8   
515935  579792  84997D     CHILDRENS CUTLERY POLKADOT PINK     8   
275327  560978   22550                    HOLIDAY FUN LUDO    12   
325850  565458   23203            JUMBO BAG VINTAGE DOILY     30   
464173  576078   84946        ANTIQUE SILVER T-LIGHT GLASS     1   
144911  548860   22997       TRAVEL CARD WALLET UNION JACK     1   
384392  570120   22570                                 NaN     6   
390592  570592   21672  WHITE SPOT RED CERAMIC DRAWER KNOB     1   

                 发票日期和时间  单价(英镑（£）)     客户编号            国家名称  
304958    8/18/2011 8:51       1.45  12415.0       Australia  
473138  11/16/2011 13:27       4.95  13258.0  United Kingdom  
163296   4/19/2011 11:37       0.83      NaN  United Kingdom  
426805  10/30/2011 13:09       1.25  14710.0  United Kingdom  
124629   3/18/2011 12:35       2.10  15019.0  United Kingdom  
255958    7/8/2011 11:36       1.25  12681.0          France  
299525   8/12/2011 10:13       1.65  13888.0  United Kingdom  
74754    1/28/2011 12:43       1.65  14659.0  United Kingdom  
479354   11/18/2011 9:57       0.85  13471.0  United Kingdom  
316247   8/30/2011 10:07       4.15  16126.0  United Kingdom  
84886     2/8/2011 12:12       4.96      NaN  United Kingdom  
32991   12/16/2010 12:31       2.95  17062.0  United Kingdom  
195741    5/19/2011 9:23       4.95  14895.0  United Kingdom  
515935  11/30/2011 15:43       4.15  12714.0          France  
275327   7/22/2011 12:20       3.75  15815.0  United Kingdom  
325850     9/5/2011 9:16       2.08  16523.0  United Kingdom  
464173  11/13/2011 16:22       1.25  14432.0  United Kingdom  
144911    4/4/2011 13:38       0.42  13263.0  United Kingdom  
384392   10/7/2011 12:49       0.00      NaN  United Kingdom  
390592  10/11/2011 11:37       2.46      NaN  United Kingdom

## 四、评估数据

#### 在这一部分，我将对在上一部分建立的 original_data 这个DataFrame所包含的数据进行评估。评估主要从两个方面进行:结构和内容，即整齐度和干净度。数据的结构性问题指不符合“每列是一个变量，每行是一个观察值每个单元格是一个值”这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。


* 1.评估数据整齐度

In [12]:
original_data.sample(10)

发票号码   产品代码                                产品名称  交易数量  \
210185  555277  22090             PAPER BUNTING RETROSPOT     3   
258133  559552  20727             LUNCH BAG  BLACK SKULL.     2   
348159  567378  22302            COFFEE MUG PEARS  DESIGN     6   
470297  576617  23378   PACK OF 12 50'S CHRISTMAS TISSUES     2   
195856  553751  84945  MULTI COLOUR SILVER T-LIGHT HOLDER    24   
394412  570872  21790                  VINTAGE SNAP CARDS     5   
533397  581023  22866       HAND WARMER SCOTTY DOG DESIGN     2   
86618   543565  22698      PINK REGENCY TEACUP AND SAUCER     6   
429305  573552  23083    SET 6 PAPER TABLE LANTERN STARS      6   
91557   544106  22766                 PHOTO FRAME CORNICE     4   

                 发票日期和时间  单价(英镑（£）)     客户编号            国家名称  
210185    6/1/2011 17:30       5.79      NaN  United Kingdom  
258133   7/10/2011 16:29       1.65  16014.0  United Kingdom  
348159   9/20/2011 10:01       2.55  14911.0            EIRE  
470297  11/15/2011 16:57       0.83      NaN  United Kingdom  
195856   5/19/2011 10:16       0.85  15630.0  United Kingdom  
394412  10/12/2011 16:40       1.63      NaN  United Kingdom  
533397   12/7/2011 10:35       4.13      NaN  United Kingdom  
86618    2/10/2011 10:40       2.95  14431.0  United Kingdom  
429305  10/31/2011 13:43       3.75  13203.0  United Kingdom  
91557    2/16/2011 10:47       2.95  14530.0  United Kingdom

从抽样的10行数据数据来看，数据符合”每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题。

* 2.评估数据的干净程度

In [15]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   发票号码       541909 non-null  object 
 1   产品代码       541909 non-null  object 
 2   产品名称       540455 non-null  object 
 3   交易数量       541909 non-null  int64  
 4   发票日期和时间    541909 non-null  object 
 5   单价(英镑（£）)  541909 non-null  float64
 6   客户编号       406829 non-null  float64
 7   国家名称       541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有541909条观察值，而Description、CustomerID 变量存在缺失值。
此外，InvoiceDate 的数据类型应为日期，CustomerID 的数据类型应为字符串，应当进行数据格式转换。

* 3.评估缺失数据

在了解产品名称存在缺失值之后，根据条件提取出缺失的每一行的观察值

In [19]:
original_data[original_data["产品名称"].isnull()]

发票号码    产品代码 产品名称  交易数量          发票日期和时间  单价(英镑（£）)  客户编号  \
622     536414   22139  NaN    56  12/1/2010 11:52        0.0   NaN   
1970    536545   21134  NaN     1  12/1/2010 14:32        0.0   NaN   
1971    536546   22145  NaN     1  12/1/2010 14:33        0.0   NaN   
1972    536547   37509  NaN     1  12/1/2010 14:33        0.0   NaN   
1987    536549  85226A  NaN     1  12/1/2010 14:34        0.0   NaN   
...        ...     ...  ...   ...              ...        ...   ...   
535322  581199   84581  NaN    -2  12/7/2011 18:26        0.0   NaN   
535326  581203   23406  NaN    15  12/7/2011 18:31        0.0   NaN   
535332  581209   21620  NaN     6  12/7/2011 18:35        0.0   NaN   
536981  581234   72817  NaN    27  12/8/2011 10:33        0.0   NaN   
538554  581408   85175  NaN    20  12/8/2011 14:06        0.0   NaN   

                  国家名称  
622     United Kingdom  
1970    United Kingdom  
1971    United Kingdom  
1972    United Kingdom  
1987    United Kingdom  
...                ...  
535322  United Kingdom  
535326  United Kingdom  
535332  United Kingdom  
536981  United Kingdom  
538554  United Kingdom  

[1454 rows x 8 columns]

从输出结果来看，这些缺失 产品名称 的交易数据，单价 都为0。为了验证猜想，我们增加筛选条件，看是否存在产品名称变量缺失目单价
不为0的数据

In [21]:
original_data[(original_data["产品名称"].isnull())&(original_data["单价(英镑（£）)"]!=0)]

Empty DataFrame
Columns: [发票号码, 产品代码, 产品名称, 交易数量, 发票日期和时间, 单价(英镑（£）), 客户编号, 国家名称]
Index: []

筛选出来结果数量为D条，说明缺失 Description 值的数据，同时也不具备有效的 UnitPrice 值。
Description 表示产品名称， Unitprice 表示产品单价，都是进行后续商品交易分析的重要变量。如果它们同时缺失/无效，我们认为数据无法提供有效含义，因此这些后续可以被删除。

In [23]:
original_data[original_data["客户编号"].isnull()]

发票号码    产品代码                             产品名称  交易数量  \
622     536414   22139                              NaN    56   
1443    536544   21773  DECORATIVE ROSE BATHROOM BOTTLE     1   
1444    536544   21774  DECORATIVE CATS BATHROOM BOTTLE     2   
1445    536544   21786               POLKADOT RAIN HAT      4   
1446    536544   21787            RAIN PONCHO RETROSPOT     2   
...        ...     ...                              ...   ...   
541536  581498  85099B          JUMBO BAG RED RETROSPOT     5   
541537  581498  85099C   JUMBO  BAG BAROQUE BLACK WHITE     4   
541538  581498   85150    LADIES & GENTLEMEN METAL SIGN     1   
541539  581498   85174                S/4 CACTI CANDLES     1   
541540  581498     DOT                   DOTCOM POSTAGE     1   

                发票日期和时间  单价(英镑（£）)  客户编号            国家名称  
622     12/1/2010 11:52       0.00   NaN  United Kingdom  
1443    12/1/2010 14:32       2.51   NaN  United Kingdom  
1444    12/1/2010 14:32       2.51   NaN  United Kingdom  
1445    12/1/2010 14:32       0.85   NaN  United Kingdom  
1446    12/1/2010 14:32       1.66   NaN  United Kingdom  
...                 ...        ...   ...             ...  
541536  12/9/2011 10:26       4.13   NaN  United Kingdom  
541537  12/9/2011 10:26       4.13   NaN  United Kingdom  
541538  12/9/2011 10:26       4.96   NaN  United Kingdom  
541539  12/9/2011 10:26      10.79   NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17   NaN  United Kingdom  

[135080 rows x 8 columns]

CustomerID 表示客户编号，不是分析畅销商品的必要变量。并且从输出结果来看，有些 CustomerID 缺失的交易数据仍然有
效，因此保留此变量为空的观察值。

* 4.评估数据重复

根据数据变量的含义来看，虽然 InvoiceNo、stockcode 和 customerID 都是唯一标识符，但一次交易可能包含多件商品，因此 InvoiceNo 可以存在重复，不同交易可以包含同件商品，因此 stockcode 可以存在重复，顾客可以进行多次交易或下单多个商品，因此 CustomerID 也可以存在重复。
那么针对此数据集，我们无需评估重复数据。

* 5.评估不一致数据

不一致数据可能存在于 Country 变量中，我们要查看是否存在多个不同值指代同一国家的情况

In [29]:
original_data["国家名称"].value_counts()

国家名称
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68
Eur

从 Country 变量值来看，"USA”、"United States"均在表示美国，"United Kingdom”、"U.K.”、"UK”均在表示英国，因此应该在清洗步骤对这些值进行统一，只保留一个指代值。

* 6.评估无效数据或错误数据

可以通过DataFrame的 describe 方法，对数值统计信息进行快速了解，

In [33]:
original_data.describe()

交易数量      单价(英镑（£）)           客户编号
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看， Quantity 和 UnitPrice 存在负数，会对后续数值分析产生影响。
因此，我们先筛选出 Quantity 数值为负数的观察值，进一步评估其含义。

In [35]:
original_data[original_data["交易数量"]<0]

发票号码    产品代码                              产品名称  交易数量  \
141     C536379       D                          Discount    -1   
154     C536383  35004C   SET OF 3 COLOURED  FLYING DUCKS    -1   
235     C536391   22556    PLASTERS IN TIN CIRCUS PARADE    -12   
236     C536391   21984  PACK OF 12 PINK PAISLEY TISSUES    -24   
237     C536391   21983  PACK OF 12 BLUE PAISLEY TISSUES    -24   
...         ...     ...                               ...   ...   
540449  C581490   23144   ZINC T-LIGHT HOLDER STARS SMALL   -11   
541541  C581499       M                            Manual    -1   
541715  C581568   21258        VICTORIAN SEWING BOX LARGE    -5   
541716  C581569   84978  HANGING HEART JAR T-LIGHT HOLDER    -1   
541717  C581569   20979     36 PENCILS TUBE RED RETROSPOT    -5   

                发票日期和时间  单价(英镑（£）)     客户编号            国家名称  
141      12/1/2010 9:41      27.50  14527.0  United Kingdom  
154      12/1/2010 9:49       4.65  15311.0  United Kingdom  
235     12/1/2010 10:24       1.65  17548.0  United Kingdom  
236     12/1/2010 10:24       0.29  17548.0  United Kingdom  
237     12/1/2010 10:24       0.29  17548.0  United Kingdom  
...                 ...        ...      ...             ...  
540449   12/9/2011 9:57       0.83  14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69  15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95  15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25  17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25  17315.0  United Kingdom  

[10624 rows x 8 columns]

从筛选结果来看， Quantity 变量为负数的观察值，InvoiceNo 似乎均以C开头，表示订单被取消。但此猜测不一定准确，为了验证猜想，我们增加筛选条件，看是否存在 Quantity 变量为负数目 InvoiceNo 不以C开头的观察值。

In [37]:
original_data[(original_data["交易数量"]<0)&(original_data["发票号码"].str[0]!="C")]

发票号码    产品代码     产品名称  交易数量          发票日期和时间  单价(英镑（£）)  客户编号  \
2406    536589   21777      NaN   -10  12/1/2010 16:50        0.0   NaN   
4347    536764  84952C      NaN   -38  12/2/2010 14:42        0.0   NaN   
7188    536996   22712      NaN   -20  12/3/2010 15:30        0.0   NaN   
7189    536997   22028      NaN   -20  12/3/2010 15:30        0.0   NaN   
7190    536998   85067      NaN    -6  12/3/2010 15:30        0.0   NaN   
...        ...     ...      ...   ...              ...        ...   ...   
535333  581210   23395    check   -26  12/7/2011 18:36        0.0   NaN   
535335  581212   22578     lost -1050  12/7/2011 18:38        0.0   NaN   
535336  581213   22576    check   -30  12/7/2011 18:38        0.0   NaN   
536908  581226   23090  missing  -338   12/8/2011 9:56        0.0   NaN   
538919  581422   23169  smashed  -235  12/8/2011 15:24        0.0   NaN   

                  国家名称  
2406    United Kingdom  
4347    United Kingdom  
7188    United Kingdom  
7189    United Kingdom  
7190    United Kingdom  
...                ...  
535333  United Kingdom  
535335  United Kingdom  
535336  United Kingdom  
536908  United Kingdom  
538919  United Kingdom  

[1336 rows x 8 columns]

以上猜想被证实错误，因为还存在 Quantity 变量为负数目 InvoiceNo 不以C开头的观察值。
但根据以上输出结果，这些筛选出的观察值的 UnitPrice 观察值均为0，因此增加 UnitPrice 的条件进行验证。

In [39]:
original_data[(original_data["交易数量"]<0)&(original_data["发票号码"].str[0]!="C")&(original_data["单价(英镑（£）)"]!=0)]

Empty DataFrame
Columns: [发票号码, 产品代码, 产品名称, 交易数量, 发票日期和时间, 单价(英镑（£）), 客户编号, 国家名称]
Index: []

根据输出结果，猜想得到验证，当 Quantity 变量为负数的时候，观察值满足以下条件之一:
* InvoiceNo 以C开头，表示订单被取消
* UnitPrice 为0，说明单价为0英镑  
这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，
因此我们将在数据清理步骤，将 Quantity 为负数的观察值删除。

In [41]:
original_data[original_data["单价(英镑（£）)"]<0]

发票号码 产品代码             产品名称  交易数量          发票日期和时间  单价(英镑（£）)  客户编号  \
299983  A563186    B  Adjust bad debt     1  8/12/2011 14:51  -11062.06   NaN   
299984  A563187    B  Adjust bad debt     1  8/12/2011 14:52  -11062.06   NaN   

                  国家名称  
299983  United Kingdom  
299984  United Kingdom

从输出结果来看， UnitPrice 为负数的观察值都是坏账调账，不属于实际商品交易数据，因此也在数据清理步骤中也将其移除。

## 清理数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括:
- 把`InvoiceDate`变量的数据类型转换为为日期时间
- 把`CustomerID`变量的数据类型转换为字符串
- 把`Description`变量缺失的观察值删除
- 把`Country`变量值`"USA"`替换为`"United states"`,把`Country`变量值`"UK"`、`"U.K."`替换为`"United Kingdom"`
- 把`Quantity`变量值为负数的观察值删除
- 把`UnitPrice` 变量值为负数的观察值删除

为了区分开经过清理的数据和原始的数据，我们创建新的变量 cleaned_data，让它为 original_data 复制出的副本。我们
之后的清理步骤都将被运用在 cleaned_data 上。

In [46]:
cleaned_data=original_data.copy()
cleaned_data.head()

发票号码    产品代码                                 产品名称  交易数量         发票日期和时间  \
0  536365  85123A   WHITE HANGING HEART T-LIGHT HOLDER     6  12/1/2010 8:26   
1  536365   71053                  WHITE METAL LANTERN     6  12/1/2010 8:26   
2  536365  84406B       CREAM CUPID HEARTS COAT HANGER     8  12/1/2010 8:26   
3  536365  84029G  KNITTED UNION FLAG HOT WATER BOTTLE     6  12/1/2010 8:26   
4  536365  84029E       RED WOOLLY HOTTIE WHITE HEART.     6  12/1/2010 8:26   

   单价(英镑（£）)     客户编号            国家名称  
0       2.55  17850.0  United Kingdom  
1       3.39  17850.0  United Kingdom  
2       2.75  17850.0  United Kingdom  
3       3.39  17850.0  United Kingdom  
4       3.39  17850.0  United Kingdom

In [47]:
cleaned_data["发票日期和时间"]=pd.to_datetime(cleaned_data["发票日期和时间"])
cleaned_data["发票日期和时间"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: 发票日期和时间, Length: 541909, dtype: datetime64[ns]

In [48]:
cleaned_data["客户编号"]=cleaned_data["客户编号"].astype(str)
cleaned_data["客户编号"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: 客户编号, Length: 541909, dtype: object

把客户编号结尾的.0去掉

In [50]:
cleaned_data["客户编号"]=cleaned_data["客户编号"].str.slice(0,-2)

In [51]:
cleaned_data["客户编号"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: 客户编号, Length: 541909, dtype: object

把Description变量缺失的观察值删除,并查看删除后该列空缺值的个数和

In [81]:
cleaned_data=cleaned_data.dropna(subset=["产品名称"])

In [83]:
cleaned_data

发票号码    产品代码                                 产品名称  交易数量  \
0       536365  85123A   WHITE HANGING HEART T-LIGHT HOLDER     6   
1       536365   71053                  WHITE METAL LANTERN     6   
2       536365  84406B       CREAM CUPID HEARTS COAT HANGER     8   
3       536365  84029G  KNITTED UNION FLAG HOT WATER BOTTLE     6   
4       536365  84029E       RED WOOLLY HOTTIE WHITE HEART.     6   
...        ...     ...                                  ...   ...   
541904  581587   22613          PACK OF 20 SPACEBOY NAPKINS    12   
541905  581587   22899         CHILDREN'S APRON DOLLY GIRL      6   
541906  581587   23254        CHILDRENS CUTLERY DOLLY GIRL      4   
541907  581587   23255      CHILDRENS CUTLERY CIRCUS PARADE     4   
541908  581587   22138        BAKING SET 9 PIECE RETROSPOT      3   

                   发票日期和时间  单价(英镑（£）)   客户编号            国家名称  
0      2010-12-01 08:26:00       2.55  17850  United Kingdom  
1      2010-12-01 08:26:00       3.39  17850  United Kingdom  
2      2010-12-01 08:26:00       2.75  17850  United Kingdom  
3      2010-12-01 08:26:00       3.39  17850  United Kingdom  
4      2010-12-01 08:26:00       3.39  17850  United Kingdom  
...                    ...        ...    ...             ...  
541904 2011-12-09 12:50:00       0.85  12680          France  
541905 2011-12-09 12:50:00       2.10  12680          France  
541906 2011-12-09 12:50:00       4.15  12680          France  
541907 2011-12-09 12:50:00       4.15  12680          France  
541908 2011-12-09 12:50:00       4.95  12680          France  

[540455 rows x 8 columns]

In [87]:
cleaned_data["产品名称"].isnull().sum()

0

把`Country`变量值`"USA"`替换为`"United states"`，并计算`USA`个数和

In [91]:
cleaned_data["国家名称"]=cleaned_data["国家名称"].replace("USA","United states")
cleaned_data["国家名称"]

0         United Kingdom
1         United Kingdom
2         United Kingdom
3         United Kingdom
4         United Kingdom
               ...      
541904            France
541905            France
541906            France
541907            France
541908            France
Name: 国家名称, Length: 540455, dtype: object

In [93]:
len(cleaned_data[cleaned_data["国家名称"]=="USA"])

0

把`Country`变量值`"UK"`、`"U.K."`替换为`"United Kingdom"`

In [98]:
cleaned_data["国家名称"]=cleaned_data["国家名称"].replace({"UK":"United Kingdom","U.K.":"United Kingdom"})

In [102]:
len(cleaned_data[cleaned_data["国家名称"]=="UK"])

0

In [104]:
len(cleaned_data[cleaned_data["国家名称"]=="U.K."])

0

把`Quantity`变量值为负数的观察值删除

In [109]:
cleaned_data=cleaned_data[cleaned_data["交易数量"]>=0]

In [113]:
len(cleaned_data[cleaned_data["交易数量"]<0])

0

把`UnitPrice` 变量值为负数的观察值删除

In [116]:
cleaned_data=cleaned_data[cleaned_data["单价(英镑（£）)"]>=0]

In [118]:
len(cleaned_data[cleaned_data["单价(英镑（£）)"]<0])

0

## 保存清理后的数据

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为e_commerce_cleaned.csv

In [121]:
cleaned_data.to_csv("e_commerce_cleaned.csv",index=False)
